packages/flutter_tools/lib/src/asset.dart

```diff
@@ -1182,6 +1175,27 @@ class ManifestAssetBundle implements AssetBundle {
     throwToolExit(errorMessage.toString());
   }
 
+  void ensureAssetPathIsValid({required String assetsBaseDir, required Uri assetUri}) {
+    if (!assetUri.isScheme('file') && assetUri.scheme.isNotEmpty) {
+      throwToolExit(
+        'Asset path "$assetUri" has scheme "${assetUri.scheme}" and is not a valid file or '
+        'directory path. Please update this entry in the pubspec.yaml to point to a valid file '
+        'path.',
+      );
+    }
+    if (Uri.directory(
+          assetsBaseDir,
+          windows: _platform.isWindows,
+        ).resolveUri(assetUri).toFilePath(windows: _platform.isWindows) ==
+        assetUri.toFilePath(windows: _platform.isWindows)) {
+      throwToolExit(
```

Comment:
I'm supportive of this change, but I do wonder if it ends up being a breaking change in practice and we could consider relaxing this (making it a warning) for a single release before making it a hard error.

On the other hand if it's very corner-casey, and we don't expect it was used much (or much intentionally), no problem doing this IMO.